pkg/xds/context/destination_index.go

```diff
@@ -0,0 +1,148 @@
+package context
+
+import (
+	"fmt"
+
+	common_api "github.com/kumahq/kuma/api/common/v1alpha1"
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core/kri"
+	"github.com/kumahq/kuma/pkg/core/resources/apis/core"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/plugins/policies/core/rules/resolve"
+)
+
+// DestinationIndex indexes destinations by KRI and labels. It provides optimized access to Kuma destinations. It should
+// be used when working with referencable destination resources like MeshServices, MeshExternalServices or MeshMultizoneServices
+type DestinationIndex struct {
+	destinationByIdentifier    map[kri.Identifier]core.Destination
+	destinationsByLabelByValue labelsToValuesToResourceIdentifier
+}
+type labelsToValuesToResourceIdentifier map[labelValue]map[kri.Identifier]bool
+
+type labelValue struct {
+	label string
+	value string
+}
+
+// DestinationResource is an interface that combines Resource with Destination. This can be used to build KRI from destinations, and access metadata of Destination
+type DestinationResource interface {
+	core_model.Resource
+	core.Destination
+}
+
+func NewDestinationIndex[T DestinationResource](destinations []T) *DestinationIndex {
+	destinationByIdentifier := make(map[kri.Identifier]core.Destination)
+	destinationsByLabelByValue := labelsToValuesToResourceIdentifier{}
+	for _, item := range destinations {
+		ri := kri.From(item, "")
+		destinationByIdentifier[ri] = item
+		buildLabelValueToServiceNames(ri, destinationsByLabelByValue, item.GetMeta().GetLabels())
```

Comment:
That's a bit of personal preference but this function is called once and something larger functions are more readable than extracting subfunctions (especially when they mutate in place).

Reply:
I personally like having these separately, it make it easier to read for me. As I said in some other comments, I like small self-contained functions that builds up to bigger functionality